crates/biome_js_analyze/src/semantic_analyzers/nursery/use_for_of.rs

```diff
@@ -0,0 +1,384 @@
+        let model = ctx.model();
+        let initializer = node.initializer()?;
+
+        if !is_initializer_valid(&initializer)? {
+            return None;
+        }
+
+        let declarators = initializer.as_js_variable_declaration()?.declarators();
+        let initializer = declarators.first()?.ok()?;
+        let initializer_id = initializer.id().ok()?;
+        let test = node.test()?;
+        let binding = initializer_id
+            .as_any_js_binding()?
+            .as_js_identifier_binding()?;
+
+        if !is_test_valid(binding, &test)? {
+            return None;
+        }
+
+        if !is_update_valid(binding, node.update()?)? {
+            return None;
+        }
+
+        let body = node.body().ok()?;
+        let block = body.as_js_block_statement();
+        let references = get_initializer_references(block, binding, model);
+        let array_right = test.as_js_binary_expression()?.right().ok()?;
+        let array_used_in_for = array_right
+            .as_js_static_member_expression()?
+            .object()
+            .ok()?;
+
+        let index_only_used_with_array = |reference| {
+            let array_in_use = reference_being_used_by_array(reference, &array_used_in_for)
+                .is_some_and(|array_in_use| array_in_use);
+            let is_delete =
+                is_delete(reference, &array_used_in_for).is_some_and(|is_assignee| is_assignee);
+
+            array_in_use && !is_delete
+        };
+
+        if references.iter().all(index_only_used_with_array) {
+            Some(())
+        } else {
+            None
+        }
+    }
+
+    fn diagnostic(node: &RuleContext<Self>, _: &Self::State) -> Option<RuleDiagnostic> {
+        Some(RuleDiagnostic::new(
+            rule_category!(),
+            node.query().range(),
+            markup! {
+                "Use "<Emphasis>"for-of"</Emphasis>" loop instead of a "<Emphasis>"for loop"</Emphasis>"."
+            },
+        ))
+    }
+}
+
+/// Get initializer references taking scope to account (excluding initials `for` loop references)
+fn get_initializer_references(
+    block: Option<&biome_js_syntax::JsBlockStatement>,
+    binding: &biome_js_syntax::JsIdentifierBinding,
+    model: &biome_js_semantic::SemanticModel,
+) -> Vec<AnyJsExpression> {
+    let skip_references = {
+        if block.is_none() {
+            2
+        } else {
+            0
+        }
+    };
+
+    binding
+        .all_references(model)
+        .skip(skip_references)
+        .filter_map(|reference| {
+            if let Some(block) = block {
+                // If block is preset, we need to consider only references in this scope
+                if block.range().start() != reference.scope().range().start() {
+                    return None;
+                }
+            }
+
+            AnyJsExpression::cast(reference.syntax().parent()?)
+        })
+        .collect()
+}
+
+/// Ensure the `for` declaration is correctly initialized
```

Comment:
This is very **generic** and the examples don't help because I can't understand **when** is valid. I don't see any syntax error, so there's some business logic that I'm not aware of.

Reply:
The code looks generic? or you are just talking about the documentation?

Reply:
I'm referring to the documentation of the function. The documentation doesn't explain when a `for` is correctly initialized. Only after 20s I understood that **in this case**, it means that it's valid only if one variable is declared.

Usually you want to *explain* in detail the requirement, then the code visually shows what was just explained

Reply:
Ok, thank you.

Can you take a look on the new one? Hope it's better now